[vcpkg-cmake] Update parallel vcpkg_cmake_configure

dg0yt opened the pull request:
- #### What does your PR fix?  
  Enables parallel cmake whenever ninja is available (non-windows hosts, system Ninja).
  Adds "Unix Makefiles" as fallback generator for non-Windows system.
  Establishs this order for generator selection:
  - If the host is x86 or `VCPKG_FORCE_SYSTEM_BINARIES` is set, try to find Ninja via `find_program`.
    - It it cannot be found, then assume that msbuild is requested for windows.
  - If msbuild is requested and suitable, use it.
  - Else if a particular generator is requested, use it.
  - Else if ninja is available, use it.
  - Else on non-windows host, use "Unix Makefiles".

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  all, no

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  incomplete

JackBoosY commented:
Should we also fix `vcpkg_configure_cmake`?

dg0yt commented:
> Should we also fix `vcpkg_configure_cmake`?

I tend to not change it:
- Ports shall be switched to `vcpkg_cmake_configure` for corrected behaviour.
- No behavioural change for building old versions of ports when they use `vcpkg_configure_cmake`.

Neumann-A commented:
i would really like separate logs for each config if it is possible.

dg0yt commented:
> i would really like separate logs for each config if it is possible.

Understood. However, this wasn't the case for windows before. 
Note that within the single log, ninja clearly separates the individual outputs. (Tested locally with port cpr.)

Neumann-A commented:
> Understood. However, this wasn't the case for windows before.
Note that within the single log, ninja clearly separates the individual outputs. (Tested locally with port cpr.)

Yeah I know. Windows always puts everything in the same file which is kind of annoying and harder to debug in my view. 
The main purpose is 
- make the files comparable to easily see differences in configure between dbg/rel
- make the errors more visible due to the separate logs

dg0yt commented:
> i would really like separate logs for each config if it is possible.

I don't see a simple way to do that portably in the ninja rule, for stdout and stderr.
This leaves only some post processing which also isn't trivial *after a failed job*.

OTOH the single config log does have some advantages:
- The number of files is reduced.
- Release config output is available even when debug config failed.

dg0yt commented:
> I don't see a simple way to do that portably in the ninja rule, for stdout and stderr. 

`cmd` has the same redirection syntax as `sh`. `cmd /C` for windows then? 
If yes, I would also move the `cmake chdir` to a plain `cd`.

dg0yt commented:
bcg729:x64-linux: source write, fix in #21516.

dg0yt commented:
> i would really like separate logs for each config if it is possible.

I'm also curious how you would think about parallelizing the build step. If this is possible at all, via `subninja`, there is no way to redirect the output. (However, I would leave the install step serialized.)

dg0yt commented:
> > I don't see a simple way to do that portably in the ninja rule, for stdout and stderr.
> 
> `cmd` has the same redirection syntax as `sh`. `cmd /C` for windows then? If yes, I would also move the `cmake chdir` to a plain `cd`.

The `cmd` approach is doable, writing the command to a `cmd` file instead of `build.ninja`. Now the problem is a different one: 
`vcpkg_execute_required_process` is unaware of the log files created via redirection. In case of error, it won't tell you to look at the split files.

dg0yt commented:
This PR is twofold, a) fixing the working directory for parallel configure, and b) applying parallel configure outside windows. Since it rebuilds a lot of ports, it should be decided if b) shall be left in this PR. 
IMO the change is okay, because the behaviour is unified, `ninja` already takes care of output order, and I prefer to see output of both build types on errors.

JackBoosY commented:
Waiting for test results.

dg0yt commented:
This is now expensively rebuilding against an old version of master. Better merge HEAD of master, to fetch the bcg729 update.
(I didn't want to trigger CI by a PR update before knowing the desired direction.)

JackBoosY commented:
@dg0yt Fine, so please merge to master to retrigger the ci test.

JackBoosY commented:
hpx regression was fixed in #21617.
python3:x64-osx:
```
install: mkdir /Users/vagrant/Data/packages/python3_x64-osx/Users/vagrant/Data/installed/x64-osx/debug/lib: File exists
make: *** [altbininstall] Error 71
```

wavelib:arm-uwp:
```
       CUSTOMBUILD : CMake error : Cannot restore timestamp "D:/buildtrees/wavelib/arm-uwp-rel/test/CMakeFiles/generate.stamp": Access is denied. [D:\buildtrees\wavelib\arm-uwp-rel\test\modwttest.vcxproj]
```

dg0yt commented:
wavelib has parallel source writes for test executables, #21671.

dg0yt commented:
> python3:x64-osx:
> 
> ```
> install: mkdir /Users/vagrant/Data/packages/python3_x64-osx/Users/vagrant/Data/installed/x64-osx/debug/lib: File exists
> make: *** [altbininstall] Error 71
> ```

In `install-x64-osx-dbg-out.log`:
~~~
...
Creating directory /Users/vagrant/Data/installed/x64-osx/debug/lib
Creating directory /Users/vagrant/Data/installed/x64-osx/debug/lib
...
~~~

This error is not related to this PR: python3 uses cmake only for windows incl. uwp. The osx build uses configure+make. 
`make altinstall` has a race condition: In `Makefile.def`, there are two subtargets which create `$(DESTDIR)$(LIBDIR)` after checking for its existence. This must be serialized.

dg0yt commented:
New issues with parallel `make altinstall` for python3:x64-osx. Why doesn't it occur outside this PR?
~~~
install: mkdir /Users/vagrant/Data/packages/python3_x64-osx/Users/vagrantinstall: mkdir /Users/vagrant/Data/packages/python3_x64-osx/Users/vagrant: File exists
: File exists
install: /Users/vagrant/Data/packages/python3_x64-osx/Users/vagrant/Data/installed/x64-osx/debug/../tools/python3/debug/bin/python3.10d: No such file or directory
make: *** [altbininstall] Error 71
make: *** Waiting for unfinished jobs....
renaming build/scripts-3.10/pydoc3 to build/scripts-3.10/pydoc3.10
renaming build/scripts-3.10/idle3 to build/scripts-3.10/idle3.10
renaming build/scripts-3.10/2to3 to build/scripts-3.10/2to3-3.10
~~~
I don't even see where it attempts to `mkdir /Users/vagrant/Data/packages/python3_x64-osx/Users/vagrant`

JackBoosY commented:
LGTM. Thanks for your effort!

dg0yt commented:
Looking at it once again:
- Maybe we need more care in adjusting `ninja_host`. At the moment it is turned off only for x86. Do we need to consider arm(64) hosts (macOS?) or BSD?
- `vcpkg_find_acquire_program(NINJA)` is called only when the generator is set to `Ninja`. But parallel configure is independent of the generator.

JackBoosY commented:
Okay, I will call my forks to review this.

BillyONeal commented:
Dropping "reviewed" not because I see problems here but because I don't have great understanding of it right now, I'll try to make sure other maintainers who are more familiar with this machinery look this week

dg0yt commented:
- vfap knows downloads for windows, darwin and linux, plus a special filesystem lookup for freebsd. It doesn't look at host arch at all.
- vfap doesn't allow failure. (IMO it should have a `REQUIRED` keyword.)

Unfortunately, this combination makes it hard to *discover* the availability of ninja. There should be only one place which implements this knowledge. Do we need port `vcpkg-ninja`?

To accelerate the fix for #21198, I could reduce the PR to this aspect, and move the host aspect to another one.

dg0yt commented:
> To accelerate the fix for #21198, I could reduce the PR to this aspect, and move the host aspect to another one.

Fix done in new PR: #22124.

BillyONeal commented:
It looks like most of the changes herein already are applied in `master`

dg0yt commented:
What is left is "Enable parallel cmake on non-windows hosts." The open questions:
- Is it really desired?
- How to detect the availability of `ninja`?

BillyONeal commented:
The maintainers talked about this in person today to answer your questions.

> * Is it really desired?

Yes, we really like this change.

> * How to detect the availability of `ninja`?

At the moment we assume Ninja is available on non-Windows hosts. Note that the existing guard for x86 is looking at Windows-specific environment variables as a workaround for lack of 32 bit builds of Ninja being available there. (We also probably have bugs on arm-windows but that's not what you're addressing here 😉)

JackBoosY commented:
Ping @dg0yt for response.

dg0yt commented:
I need to revise the changes.

dg0yt commented:
Error classes (from https://dev.azure.com/vcpkg/public/_build/results?buildId=69374&view=results)
- Forbidden source writes: brpc, qtdeviceutilities
- Needs to find `ninja` executable: qtwebengine (custom FindNinja.cmake), arrow, clblas (`ExternalProject.cmake`, https://gitlab.kitware.com/cmake/cmake/-/issues/23355) 
- Baseline regressions: qtinterfaceframework (#23766), ezc3d (#23781), qtdeclarative (#23746)
- Supposed unrelated race in msbuild: nanoflann:arm-uwp

dg0yt committed:
Revise generator selection

If the host is x86, assume that msbuild is requested for windows.
If msbuild is requested and suitable, use it.
Else if a particular generator is requested, use it.
Else if ninja is available, use it.
Else on non-windows host, use "Unix Makefiles".

dg0yt committed:
Consolidate ninja path setup

dg0yt committed:
Revise ninja_host detection

dg0yt committed:
Revise parallel configure detection

dg0yt committed:
Update documentation

dg0yt commented on ports/vcpkg-cmake/vcpkg_cmake_configure.cmake:
```diff
@@ -147,22 +148,26 @@ function(vcpkg_cmake_configure)
     if(DEFINED arg_GENERATOR)
         set(generator "${arg_GENERATOR}")
-    elseif(arg_WINDOWS_USE_MSBUILD OR NOT ninja_can_be_used)
+    elseif(arg_WINDOWS_USE_MSBUILD AND VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_TARGET_IS_MINGW)
```
TBH I would prefer to move this before the `arg_GENERATOR` branch. This would allow to use `GENERATOR "Unix Makefiles" WINDOWS_USE_MSBUILD` with a reasonable effect.

dg0yt commented:
I'm going to continue when vcpkg-cmake dust has settled. Three more PRs against this port, and no coordination,

JackBoosY commented:
Closing this PR since it seems that no progress is being made. Please reopen if work is still being done.

dg0yt commented:
I cannot reopen. Stop closing my drafts.

dg0yt committed:
Use portable chdir

dg0yt committed:
[ms-gltf] Use new generator selection

dg0yt commented:
I will update the version database when the PR received preliminary acceptance.

cenit commented on ports/ms-gltf/vcpkg.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "ms-gltf",
-  "version-date": "2022-06-28",
+  "version-date": "2022-07-18",
```
should bump port version, not version-date

BillyONeal replied:
If the content of the stuff we install changed, then the version-date should be updated. If only maintainer improvement stuff in the port changed, port-version should be bumped.

dg0yt commented:
It is nice to know that there is a micro issue with the ms-gltf version update. But before burning more CI time, I need feedback on the core part.

dg0yt committed:
Update versions

BillyONeal reviewed (approved):
I note that there is a preexisting issue related to vcpkg_find_acquire_program and FORCE_SYSTEM_BINARIES I did not comment on here since it's unrelated to this improvement. (And I don't think other reviewers should block for that either)

BillyONeal commented:
Thanks for making parallelism work in more places:)

dg0yt commented:
Now I may turn to `vpckg_configure_make`. Which has quite some potential.

Neumann-A commented:
> Now I may turn to vpckg_configure_make. Which has quite some potential.

do/test it outside of that function ;) Make a script port for it.
